### PR TITLE
feat(cache): add ZCache timestamp-indexed message cache

### PR DESCRIPTION
**Pull request by YuanYuYuan**
## Summary

Adds `ZCache<T>`, a timestamp-indexed sliding-window message cache that mirrors the ROS 2 `message_filters::Cache<T>` API. It wraps a Zenoh subscriber and retains up to N messages queryable by time, without requiring a background thread or executor.

## Key Changes

- `crates/ros-z/src/cache.rs`: new `ZCache<T>`, `ZCacheBuilder`, `CacheInner`, `ZenohStamp`, `ExtractorStamp` types
- `ZNode::create_cache::<T>(topic, capacity)` factory method with type-state builder
- Two stamp strategies selected at build time: `ZenohStamp` (Zenoh transport timestamp, zero-config) and `ExtractorStamp` (user-supplied closure, e.g. `header.stamp`)
- Query API: `get_interval`, `get_before`, `get_after`, `get_nearest`, `oldest_stamp`, `newest_stamp`, `len`, `is_empty`, `clear`
- `get_interval` with inverted range returns empty instead of panicking
- `get_nearest` tie-breaking: equidistant messages prefer the earlier timestamp
- `ZCache`, `ZCacheBuilder`, `ZenohStamp`, `ExtractorStamp` re-exported from `prelude`
- 19 integration tests covering query semantics, boundary conditions, capacity eviction, collision handling, concurrent access, and subscriber lifecycle
- `examples/z_cache.rs`: runnable demo with both stamp modes and CLI flags
- `book/src/chapters/message_cache.md`: new book chapter covering both stamp strategies, full query API, capacity semantics, and a comparison table with `message_filters::Cache<T>`

## Breaking Changes

None

### CODE REVIEW

File: crates/ros-z/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+    let session = sub_builder.session.clone();
+    let keyexpr_format = sub_builder.keyexpr_format;
+    let entity = sub_builder.entity.clone();
+
+    let topic_ke = keyexpr_format.topic_key_expr(&entity)?;
+    let key_expr = (*topic_ke).clone();
+    debug!("[CACHE] Key expression: {}", key_expr);
+
+    let inner = Arc::new(Mutex::new(CacheInner::<T>::new(capacity)));
+    let inner_cb = Arc::clone(&inner);
```
Comment (schmidma) [inline crates/ros-z/src/cache.rs]: is there a reason not to use
```suggestion
    let inner_cb = inner.clone();
```
